File: logparse.pl
#!/usr/bin/env perl
# $Header:  logparse.pl v.0.1.2 2014/06/16    al.netrusov@gmail.com   $
# Author:   Alexander Netrusov
###################################################################################
use strict;
use warnings;
use utf8;

use English qw(-no_match_vars);
use Carp ();

use FindBin               ();
use Digest::MD5           ();
use File::Spec::Functions ();

Carp::confess
  "Usage: ${PROGRAM_NAME} <file>[+multiline] <mask> <match> [<except>]"
  if !@ARGV || scalar @ARGV < 3;

my $argv = {};

for (qw|file mask match except|) {
    $argv->{$_} = shift @ARGV;
}

my ( $file, $multiline ) = $argv->{file} =~ m/(.+?)([+]multi[line]{0,4})?$/xms;

open my $fh, '<', $file
  or Carp::confess sprintf "Failed to open '%s': %s", $file,
  $EXTENDED_OS_ERROR;

seek $fh, readpos($file), 0;

my $entry = {};

while (<$fh>) {
    undef $entry->{final};

    if (m/$argv->{mask}/xms) {
        $entry->{final} = $entry->{temp};
        $entry->{temp}  = $_;
    }
    else {
        $multiline ? $entry->{temp} .= $_ : undef $entry->{temp};
    }

    for ( eof $fh ? ( 'final', 'temp' ) : 'final' ) {
        if ( $entry->{$_} && $entry->{$_} =~ m/$argv->{match}/xms ) {
            next
              if $argv->{except}
              && $entry->{$_} =~ m/$argv->{except}/xms;

            print $entry->{$_};
        }
    }
}

writepos( $file, tell $fh );

close $fh
  or Carp::confess 'Failed to close filehandle: ' . $EXTENDED_OS_ERROR;

sub readpos {
    my ($file) = @_;

    $file = File::Spec::Functions::catfile( $FindBin::Bin,
        q{.} . Digest::MD5::md5_hex($file) );

    my $pos = 0;

    if ( -f $file ) {
        open my $fh, '<', $file
          or Carp::confess "Failed to open '$file': " . $EXTENDED_OS_ERROR;

        chomp( $pos = <$fh> );

        close $fh
          or Carp::confess 'Failed to close filehandle: ' . $EXTENDED_OS_ERROR;
    }

    return $pos;
}

sub writepos {
    my ( $file, $pos ) = @_;

    $file = File::Spec::Functions::catfile( $FindBin::Bin,
        q{.} . Digest::MD5::md5_hex($file) );

    open my $fh, '>', $file
      or Carp::confess sprintf "Failed to open '%s': %s", $file,
      $EXTENDED_OS_ERROR;

    print {$fh} $pos
      or Carp::confess sprintf "Failed to write to '%s': %s", $file,
      $EXTENDED_OS_ERROR;

    close $fh
      or Carp::confess 'Failed to close filehandle: ' . $EXTENDED_OS_ERROR;

    return 1;
}

__END__

File: logparse.py
import sys
import os
import re
import hashlib

if len(sys.argv) < 4:
    print('Usage: logparse.py <file>[+multiline] <mask> <match> [<except>]')
    sys.exit(1)

argv = {}

for arg in 'file', 'mask+re', 'match+re', 'except+re':
    arg, regex = re.match(r'^(.+?)([+]re)?$', arg).groups()
    try:
        argv[arg] = sys.argv.pop(1)
        if regex:
            argv[arg] = re.compile(argv[arg])
    except IndexError:
        argv[arg] = None

argv['file'], multiline = re.match(r'^(.+?)([+]multi[line]{0,4})?$', argv['file']).groups()

entry = {
    'final': None,
    'temp': None
}


def readpos(file):
    file = os.path.join(sys.path[0], '.' + hashlib.md5(bytes(file, encoding='utf-8')).hexdigest())

    try:
        with open(file, 'r') as f:
            position = f.readline()
    except FileNotFoundError:
        position = 0

    return int(position)


def writepos(file, position):
    file = os.path.join(sys.path[0], '.' + hashlib.md5(bytes(file, encoding='utf-8')).hexdigest())

    with open(file, 'w') as f:
        f.write(str(position))


with open(argv['file'], 'r') as f:
    f.seek(readpos(argv['file']))

    while True:
        line = f.readline()

        entry['final'] = None

        if argv['mask'].search(line):
            entry['final'] = entry['temp']
            entry['temp'] = line
        else:
            if multiline:
                entry['temp'] += line

        for _entry in ['final'] if line else ['final', 'temp']:
            if entry[_entry]:
                if argv['match'].search(entry[_entry]):
                    if argv['except'] and argv['except'].search(entry[_entry]):
                        pass
                    else:
                        print(entry[_entry])
        if not line:
            break

    writepos(argv['file'], f.tell())
